src/Utilities/ProcessExtensions.cs

```diff
@@ -0,0 +1,122 @@
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using Microsoft.Build.Shared;
+
+namespace Microsoft.Build.Utilities
+{
+    internal static class ProcessExtensions
+    {
+#if !RUNTIME_TYPE_NETCORE
+        private static readonly bool _isWindows = true;
+#else
+        private static readonly bool _isWindows = RuntimeInformation.IsOSPlatform(OSPlatform.Windows);
+#endif
+
+        public static void KillTree(this Process process, int timeout)
+        {
+            string stdout;
+            if (_isWindows)
+            {
+                try
+                {
+                    // issue the kill command
+                    NativeMethodsShared.KillTree(process.Id);
+                }
+                catch (InvalidOperationException)
+                {
+                    // The process already exited, which is fine,
+                    // just continue.
+                }
+
+                // wait until the process finishes exiting/getting killed. 
+                // We don't want to wait forever here because the task is already supposed to be dieing, we just want to give it long enough
+                // to try and flush what it can and stop. If it cannot do that in a reasonable time frame then we will just ignore it.
+                process.WaitForExit(timeout);
+            }
+            else
+            {
+                var children = new HashSet<int>();
+                GetAllChildIdsUnix(process.Id, children, timeout);
+                foreach (var childId in children)
+                {
+                    KillProcessUnix(childId, timeout);
+                }
+                KillProcessUnix(process.Id, timeout);
+            }
+        }
+
+        private static void GetAllChildIdsUnix(int parentId, ISet<int> children, int timeout)
+        {
+            string stdout;
+            var exitCode = RunProcessAndWaitForExit(
+                "pgrep",
+                $"-P {parentId}",
+                timeout,
+                out stdout);
+
+            if (exitCode == 0 && !string.IsNullOrEmpty(stdout))
+            {
+                using (var reader = new StringReader(stdout))
+                {
+                    while (true)
+                    {
+                        var text = reader.ReadLine();
+                        if (text == null)
+                        {
+                            return;
+                        }
+
+                        int id;
+                        if (int.TryParse(text, out id))
+                        {
+                            children.Add(id);
+                            // Recursively get the children
+                            GetAllChildIdsUnix(id, children, timeout);
+                        }
+                    }
+                }
+            }
+        }
+
+        private static void KillProcessUnix(int processId, int timeout)
+        {
+            string stdout;
+            RunProcessAndWaitForExit(
```

Comment:
I have changed it so the timeout will only be used to wait for this process.
Regarding the issuing of signal(3) vs. shelling out the kill command, IMO, this is a port of what CLI is using to unblock couple of issues. We can latest evaluate whether it makes sense to P/Invoke syscall and how much beneficial would that actually be in real world example.